### ISSUE TITLE
sparse + dask array not supported

### ISSUE THREAD

**Issue by shoyer** [dask/dask#5259]
cc @hameerabbasi 

dask+sparse works, but not sparse+dask.

I think dask should probably implement this case, via it's `__array_ufunc__` method. Here's the traceback:
```
In [1]: import sparse

In [2]: import dask.array as da

In [3]: import numpy as np

In [4]: x = np.arange(3)

In [5]: y = sparse.COO(x)

In [6]: z = da.from_array(y, chunks=3)

In [7]: z + y
Out[7]: dask.array<add, shape=(3,), dtype=int64, chunksize=(3,)>

In [8]: y + z
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-8-ff96b8185f8d> in <module>
----> 1 y + z

~/miniconda3/envs/xarray-py37-dev/lib/python3.7/site-packages/numpy/lib/mixins.py in func(self, other)
     23         if _disables_array_ufunc(other):
     24             return NotImplemented
---> 25         return ufunc(self, other)
     26     func.__name__ = '__{}__'.format(name)
     27     return func

TypeError: operand type(s) all returned NotImplemented from __array_ufunc__(<ufunc 'add'>, '__call__', <COO: shape=(3,), dtype=int64, nnz=2, fill_value=0>, dask.array<array, shape=(3,), dtype=int64, chunksize=(3,)>): 'COO', 'Array'
```

**Comment by mrocklin** [comment]
What do you think would be the right approach here?  Call `da.from_array` on all non-dask arrays?

Currently we have this code, which seems problematic: 

https://github.com/dask/dask/blob/6f382404707df2d7ca70d44bd36ecae485de4028/dask/array/core.py#L1106-L1110

I wonder if just removing that check would make things work

**Comment by hameerabbasi** [comment]
I think probably adding a `hasattr(..., '__array_ufunc__') would help here.

**Comment by shoyer** [comment]
Xarary defines __array_ufunc__, so that wouldn't be enough. Dask needs to
defer to xarray, but handle sparse arrays itself. I don't know if there's a
good way to do that without an explicit list of valid or invalid "array
chunk types".

**Comment by hameerabbasi** [comment]
Maybe there needs to be a way to declare something as “ndarray-compatible”, this has come up before.

**Comment by shoyer** [comment]
> Maybe there needs to be a way to declare something as “ndarray-compatible”, this has come up before.

Right, but remember that other "ndarray-compatible" arrays might need to wrap dask arrays. For example, if someone wrote a duck type compatible version of autograd.

**Comment by jakirkham** [comment]
> What do you think would be the right approach here? Call `da.from_array` on all non-dask arrays?

We call `da.asarray` in other places to handle non-Dask Arrays. Seems reasonable to do it here too.

**Comment by shoyer** [comment]
> We call `da.asarray` in other places to handle non-Dask Arrays. Seems reasonable to do it here too.

yes, though with overriding numpy ufuncs we still have to be careful to explicitly skip types like xarray.DataArray

**Comment by hameerabbasi** [comment]
> Right, but remember that other "ndarray-compatible" arrays might need to wrap dask arrays. For example, if someone wrote a duck type compatible version of autograd.

Wouldn’t autograd be wrapped by Dask Array instead?

**Comment by jakirkham** [comment]
> > We call `da.asarray` in other places to handle non-Dask Arrays. Seems reasonable to do it here too.
> 
> yes, though with overriding numpy ufuncs we still have to be careful to explicitly skip types like xarray.DataArray

This is what I'm hoping the `asduckarray` function will do for us in libraries like Dask. 😉

xref: https://github.com/numpy/numpy/issues/13831 (for others interested in learning more)
